### ISSUE TITLE
is_key_down keep being true when window gets unfocused

### ISSUE THREAD

**Issue by lomstfer** [not-fl3/macroquad#547]
1. I press and hold down a key.
2. I make the window not in focus (on windows) and release the key.
3. I see that is_key_down of the key that I pressed is still true, until i focus the window again and release the key again.

**Comment by miwarnec** [comment]
This happens to me as well for macroquad 0.3.X.
For example, on Windows: press Windows key.
- Key pressed fires once
- Key down keeps firing while in background since the start menu is now in foreground
- Switch back to macroquad application, key down still keeps firing

in case of the Windows key, this can't even be released since pressing it again opens the windows start menu again.

**Comment by profan** [comment]
Definitely also still happening on macroquad 0.4.x for me, releasing keys with the mouse off the window means you don't seem to get notified of the input at all, which causes a lot of slightly funky behaviour if you happen to accidentally mouse off the window momentarily (while for instance, middle mouse panning or similar).

**Comment by jangler** [comment]
Has anyone figured out a workaround for this?

**Comment by profan** [comment]
@jangler god I wish 😢 

> Has anyone figured out a workaround for this?

god I wish, if you do let me know because it's a royal pain for sure right now

**Comment by jangler** [comment]
In my use case I'm mostly concerned with predictable loss of focus from using `rfd`, so the workaround would be either:

1. Track input state yourself (using `get_keys_pressed`, etc.) and clear it when opening a file dialog
2. Patch in a method to reset the input state of macroquad's context object

But of course this is useless for scenarios where focus is lost unpredictably.

**Comment by chamons** [comment]
Looking into this, I think miniquad would need a window focus event (see note on https://docs.rs/miniquad/latest/miniquad/window/index.html) for macroquad to listen to and reset pressed keys?

**Comment by jangler** [comment]
That's my understanding!

**Comment by profan** [comment]
It looks like there's been a bandaid fix for this, but only for wayland in: https://github.com/not-fl3/miniquad/pull/519

Fairly confident this is still an issue even after that PR, at least definitely on non-linux platforms and just clearing the modifiers doesn't feel like the right fix anyways

**Comment by jangler** [comment]
In my experience the pressed/released state also needs to be cleared.

**Comment by jangler** [comment]
Oh, I guess my issue was specific to blocking dialogs then!

**Comment by jonatino** [comment]
Might try and fix this soon. Is there any situation where you'd still want to accept user input when a window is not in focus? I can't think of any scenarios with touch/keyboard/mouse input where you'd want to accept input when not in focus, but what about if a controller is plugged in? Still ignore all input until window is back in focus?

**Comment by profan** [comment]
@jonatino I feel like if a key/mouse button/etc was pressed while inside the window, and you unfocus/mouse off the window, you'd still want to receive the released event for those events at least

**Comment by jangler** [comment]
Clearing all the pressed keys/buttons when focus is lost is the usual behavior, though.

**Comment by jonatino** [comment]
@jangler yeah I tend to agree with you. I don't think what @profan wants is possible across all operating systems. My solution was simply to clear all input state on window defocus. The sticky key issue you have will likely be fixed by this since you'd know right away after the window loses focus that the key is no longer held.

**Comment by profan** [comment]
@jangler I don't think you'd get that on SDL by default or even just with Godot, but as long as the sticky key issue is resolved somehow I'll be placated as at least now the game will at least feel less broken :)

**Comment by jonatino** [comment]
@profan what you're asking for doesn't exist in SDL and is impossible on half the platforms that macroquad supports so it will lead to inconsistent behavior in the engine which I imagine @not-fl3 doesn't want. 

Eg, SDL you need to specify u want to handle gamecontroller inputs when window is unfocused. By default, idk of any operating system which forwards input events to windows that arent in focus. This is done on purpose as a security reason otherwise any app on your computer could keylog you with no special permissions (since they'd get input events even when not in focus). The only way would to get input events out of focus would be raw input hooking but that impossible todo on web and mobile without permissions that most people wont grant for a game.

https://wiki.libsdl.org/SDL3/SDL_HINT_JOYSTICK_ALLOW_BACKGROUND_EVENTS

**Comment by profan** [comment]
@jonatino Right, well whatever the outcome is, as long as it results in a consistent observable behaviour for miniquad and macroquad applications where one half of an input event doesn't go missing (ie. any key pressed has an accompnaying key released, or you end up with inconsistent game states or having to deal with the fact that you might never get a key released for a given key pressed)

**Comment by InnocentusLime** [comment]
> [@jonatino](https://github.com/jonatino) Right, well whatever the outcome is, as long as it results in a consistent observable behaviour for miniquad and macroquad applications where one half of an input event doesn't go missing (ie. any key pressed has an accompanying key released, or you end up with inconsistent game states or having to deal with the fact that you might never get a key released for a given key pressed).
> 
> I get your point about window focus and that generally unfocused windows don't usually get input events for obvious security reasons (and also it's probably just a mess if all applications would be capturing input events).
> 
> Overall I don't really care that much what we end up with, I just want the observable behaviour to not a little bit of a mess as it currently is (especially as you barely have window focus information in macroquad applications, so fixing the pressed/released state stuff is harder for someone who isn't working at the miniquad level).
> 
> Example of how Godot sorted this out: [godotengine/godot#28061](https://github.com/godotengine/godot/pull/28061)

I decided to give that issue a shot :)
It seems we would need both `miniquad` and `macroquad` patches. The `miniquad` patch can be found here: `not-fl3/miniquad#552`.
I will test it on OsX a bit later, but if someone else tests it on their machines -- it would be greatly appreciated.

**Comment by jonatino** [comment]
@InnocentusLime I also started on this, just threw something together in the past 10mins.

Added a focus gained/lost event for the window.

Miniquad commits:
windows & wasm: https://github.com/VoltaSoftware/miniquad/commit/59549b66aba7368a06731c4b901846983dabd5c6
android & linux: https://github.com/VoltaSoftware/miniquad/commit/5a87f41b5d0537c9a312511a9b209405c6481342
macos: https://github.com/VoltaSoftware/miniquad/commit/3e3263384a4975b51e6eabbcb79b89bdd7aea0e0
ios: https://github.com/VoltaSoftware/miniquad/commit/9a8af70ea3f3f37fbbd86ee81e84aca9f496cb45

Macroquad commit:
https://github.com/VoltaSoftware/macroquad/commit/49972d9a16ce7b19cb08526ac16362138af70791?w=1#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759
